### PR TITLE
PETScWrappers:BlockSparseMatrix: create PETSc MATNEST

### PR DESCRIPTION
**Pull request by luca-heltai**
This is the first step towards using PETSc solvers with our BlockMatrix wrappers.

### CODE REVIEW

File: doc/news/changes/minor/20221220StefanoZampini-b [hunk 1]
```diff
@@ -0,0 +1,3 @@
+New: PETScWrappers:BlockSparseMatrix now is also a PETSc MATNEST type.
+<br>
+(StefanZampini, 2022/12/20)
```
Comment (drwells) [inline doc/news/changes/minor/20221220StefanoZampini-b:3]:
```suggestion
(Stefano Zampini, 2022/12/20)
```
:)

File: include/deal.II/lac/petsc_block_sparse_matrix.h [hunk 2]
```diff
@@ -271,6 +271,26 @@ namespace PETScWrappers
...
       using BlockMatrixBase<SparseMatrix>::clear;
+
+      /**
+       * Conversion operator to gain access to the underlying PETSc type. If you
+       * do this, you cut this class off some information it may need, so this
+       * conversion operator should only be used if you know what you do. In
+       * particular, it should only be used for read-only operations into the
+       * matrix.
+       */
+      operator Mat() const;
```
Comment (drwells) [inline include/deal.II/lac/petsc_block_sparse_matrix.h:282]:
```suggestion
      operator const Mat &() const;
```
Lets make this consistent with the vector class.

File: source/lac/petsc_parallel_block_sparse_matrix.cc [hunk 3]
```diff
@@ -31,6 +32,11 @@ namespace PETScWrappers
       return *this;
     }
 
+    BlockSparseMatrix::~BlockSparseMatrix()
+    {
+      PetscErrorCode ierr = destroy_matrix(petsc_nest_matrix);
+      AssertThrow(ierr == 0, ExcPETScError(ierr));
```
Comment (drwells) [inline source/lac/petsc_parallel_block_sparse_matrix.cc:38]:
```suggestion
      AssertNothrow(ierr == 0, ExcPETScError(ierr));
```
No throwing from dtors

FYI: like the other PR: we do have `destroy_matrix()` but it doesn't do anything but call `MatDestroy()` these days. It was written at a time when we supported old versions of PETSc which had a different API. I'd prefer to just remove the function and call `MatDestroy()` directly.

File: source/lac/petsc_parallel_block_sparse_matrix.cc [hunk 4]
```diff
@@ -111,6 +117,24 @@ namespace PETScWrappers
...
+        for (unsigned int c = 0; c < n; c++)
+          psub_objects[r * n + c] = this->sub_objects[r][c]->petsc_matrix();
+      ierr = MatCreateNest(get_mpi_communicator(),
+                           m,
+                           NULL,
+                           n,
+                           NULL,
+                           psub_objects.data(),
+                           &petsc_nest_matrix);
+      AssertThrow(ierr == 0, ExcPETScError(ierr));
```
Comment (drwells) [inline source/lac/petsc_parallel_block_sparse_matrix.cc:137]: ✔️

**Review by drwells — CHANGES_REQUESTED** [review]

**Comment by drwells** [comment]
This looks good. I wrote a much worse version of this in September to make step-32 work with PETSc.

**Review by drwells — APPROVED** [review]

**Review by stefanozampini — APPROVED** [review]

**Comment by luca-heltai** [comment]
CI fails when installing mold linker.

**Comment by luca-heltai** [comment]
@drwells, shall we merge?

**Pull request merged**